torch/_decomp/decompositions.py

```diff
@@ -641,7 +641,10 @@ def masked_fill_Scalar(self: Tensor, mask: Tensor, value: float) -> Tensor:
 
 @register_decomposition(aten.masked_fill.Tensor)
 def masked_fill_Tensor(self: Tensor, mask: Tensor, value: Tensor) -> Tensor:
-    return torch.where(mask, value, self)
+    # `masked_fill` allows cpu scalar to be moved to cuda but not otherwise.
+    if self.device.type == "cuda" and value.device.type == "cpu":
+        value = value.to(self.device)
+    return torch.where(mask, value.to(self.dtype), self)
```

Comment:
```suggestion
    value = value.to(self.dtype)
    if self.device.type == "cuda" and value.device.type == "cpu":
        value = value.item()
    return torch.where(mask, value, self)
```
avoid syncronizing `to`